apps/wing/src/commands/test/snapshots.ts

```diff
@@ -0,0 +1,178 @@
+import * as cp from "child_process";
+import { existsSync, readFileSync, writeFileSync } from "fs";
+import * as fs from "fs/promises";
+import { join, extname, basename } from "path";
+import { BuiltinPlatform } from "@winglang/compiler";
+import * as glob from "glob";
+import { TestOptions } from "./test";
+import { renderTestName } from "./util";
+import { withSpinner } from "../../util";
+import { compile } from "../compile";
+import { SNAPSHOTS_ERROR_HELP } from "./snapshots-help";
+
+export enum SnapshotMode {
+  /**
+   * Auto-detect. If CI=1, SnapshotMode.ASSERT otherwise SnapshotMode.DEPLOY.
+   */
+  AUTO = "auto",
+
+  /**
+   * Snapshots are disabled.
+   */
+  NEVER = "never",
+
+  /**
+   * Deploy and test in the cloud and update the snapshot only if all tests passed.
+   */
+  DEPLOY = "deploy",
+
+  /**
+   * Update the snapshots without deploying (dry run).
+   */
+  UPDATE = "update",
+
+  /**
+   * Just verify that the snapshots are correct.
+   */
+  ASSERT = "assert",
+}
+
+export function determineSnapshotMode(
+  target: string | undefined,
+  options: TestOptions
+): SnapshotMode {
+  // no snapshots for sim, ever!
+  if (target === BuiltinPlatform.SIM) {
+    return SnapshotMode.NEVER;
+  }
+
+  // if AUTO or unset, determine behavior based on CI flag
+  if (!options.snapshots || options.snapshots === SnapshotMode.AUTO) {
+    // determine behavior based on CI
+    if (process.env.CI) {
+      return SnapshotMode.ASSERT;
+    } else {
+      return SnapshotMode.DEPLOY;
+    }
+  }
+
+  return options.snapshots;
+}
+
+export async function captureSnapshot(entrypoint: string, target: string, options: TestOptions) {
+  const snapshotMode = determineSnapshotMode(target, options);
+
+  // skip if snapshots are disabled
+  if (snapshotMode === SnapshotMode.NEVER) {
+    return;
+  }
+
+  const snapshotFile = `${entrypoint}.${target}.snap.md`;
+
+  await withSpinner(`Snapshot ${renderTestName(snapshotFile)}...`, async () => {
+    // we need to compile again in order for snapshots because we can't afford the snapshot to be
+    // based on a random root id (which is the default for tests)
+    const snapshotDir = await compile(entrypoint, options);
+
+    // take a snapshot of the synthesis output
+    const snapshot = await createMarkdownSnapshot(basename(snapshotFile), snapshotDir);
+
+    // update the snapshots if needed
+    switch (snapshotMode) {
+      case SnapshotMode.DEPLOY:
+      case SnapshotMode.UPDATE:
+        writeFileSync(snapshotFile, snapshot);
+        break;
```

Comment:
Can we check the contents of the snapshot to see if it actually changed and avoid writing to it if not.

And if we do write to it, it'd be useful to let the user know as well.

Reply:
Ok